metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+        return dict_.lookup(string_kmer, check_reverse_complement);
+    }
+
+    uint64_t lookup_uint(kmer_t uint_kmer, bool check_reverse_complement = true) const {
+        return dict_.lookup_uint(uint_kmer, check_reverse_complement);
+    }
+
+    sshash::lookup_result lookup_advanced(
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.lookup_advanced(string_kmer, check_reverse_complement);
+    }
+
+    sshash::lookup_result lookup_advanced_uint(
+            kmer_t uint_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.lookup_advanced_uint(uint_kmer, check_reverse_complement);
+    }
+
+    sshash::neighbourhood<kmer_t> kmer_forward_neighbours(
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.kmer_forward_neighbours(string_kmer, check_reverse_complement);
+    }
+
+    sshash::neighbourhood<kmer_t> kmer_backward_neighbours(
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.kmer_backward_neighbours(string_kmer, check_reverse_complement);
+    }
+
+  private:
+    // TODO: this is a hack since sshash::dictionary<kmer_t>::visit is always non-const
+    mutable sshash::dictionary<kmer_t> dict_;
+};
+
+constexpr DeBruijnGraph::node_index sshash_to_graph_index(uint64_t idx) {
+    return idx + 1;
+}
+constexpr uint64_t graph_index_to_sshash(DeBruijnGraph::node_index idx) {
+    return idx - 1;
+}
+
+DBGSSHash::DBGSSHash(size_t k, Mode mode) : k_(k), num_nodes_(0), mode_(mode) {
+    size_t odd_k = k_ % 2 == 0 ? k_ + 1 : k_;
+
+    if (odd_k * IDictionary::bits_per_char <= 64) {
+        dict_ = std::make_unique<Dictionary<Kmer64>>();
+    } else if (odd_k * IDictionary::bits_per_char <= 128) {
+        dict_ = std::make_unique<Dictionary<Kmer128>>();
+    } else if (odd_k * IDictionary::bits_per_char <= 256) {
+        dict_ = std::make_unique<Dictionary<Kmer256>>();
+    } else {
+        common::logger->error("k too big: {} > {}", odd_k, 256 / IDictionary::bits_per_char);
+        throw std::runtime_error("k fail");
+    }
+}
+
+DBGSSHash::DBGSSHash(std::string const& input_filename, size_t k, Mode mode)
+      : DBGSSHash(k, mode) {
+    sshash::build_configuration build_config;
+    build_config.k = k;
+    // quick fix for value of m... k/2 but odd
+    build_config.m = std::min(uint64_t((k_ + 1) / 2), sshash::constants::max_m);
+    build_config.verbose = common::get_verbose();
+    build_config.num_threads = get_num_threads();
+    if (build_config.m % 2 == 0)
+        build_config.m++;
+
+    // silence sshash construction messages when not verbose
+    if (!common::get_verbose())
+        std::cout.setstate(std::ios_base::failbit);
+
+    dict_->build(input_filename, build_config);
+    if (!common::get_verbose())
+        std::cout.clear();
```

Comment:
Eh, somewhat hacky. Generally I'd be wary of doing this because what if e.g. some third party uses metagraph and wants to silence something too, and we would unsilence it prematurely? I feel like
```cpp
std::ostringstream local;
auto cout_buff = std::cout.rdbuf();
std::cout.rdbuf(local.rdbuf());

/* call library function here !!! */

std::cout.rdbuf(cout_buff);
```
from [here](https://cplusplus.com/forum/general/283883/#msg1229098) is a better way of silencing, because it would at least restore `std::cout` to exactly the state it was before the call. Welp, hope we won't get into multithreading issues while doing so...

Reply:
sshash actually does have a verbose mode, but they unfortunately don't have a quiet mode...the ideal thing would be for us to add it

Reply:
Let's try rdbuf swap for now? And we can open an issue in sshash upstream requesting a silent mode.

Reply:
I'm afraid of multithreading issues, to be honest, that's why I didn't use the approach you're suggesting. But I agree that the one I did is not ideal either...

Reply:
Is there any reason to believe that the current approach is more thread-safe than rdbuf swap? As the link says, 
> **Any** "hack" that temporarily disables/redirects std::cout around library calls is not thread-safe.

Reply:
True...it was more of a hunch that changing the status of cout would be more thread safe than changing the rdbuf, but I have no actual proof of that.

Reply:
I think I'd still try to use rdbuf for now then. I feel like messing with global state without properly reverting it back is a somewhat more unwanted and unexpected behavior for third-party users.  We can then reconsider if any actual issues occur?

Reply:
Sounds good

Reply:
Changed to rdbuf swap. Let's see how it behaves now.